src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -266,8 +299,8 @@ protected virtual void Lock()
 			{
 				if (!_needCompletionLocking || _isDisposed)
 					return;
-				_needCompletionLocking = false;
 				_lock.Reset();
+				_needCompletionLocking = false;
```

Comment:
There was a concurrency issue possible here, allowing the session disposal to not wait where it should have. But the case I have seen may be bound to the theoretical case for which I have put initially a crutch.

I will try for a few more days to reproduce the issue with this change, to somewhat check if the session not closed issue does still occur.

Reply:
The failure does still happen. I have put back in place the crutch, but only for the case which can fail : when session usage from transaction scope completion is not disabled. That is a legacy case we advise to not use.

I will again try for a few days to reproduce the issue.

Reply:
Reverting: the unclosed session issue does still occur. So, there is a remaining concurrency issue, but only for the case where using the session from transaction completion is not disabled as recommended. I think I let it lie there.